properties/extras/uid.py

```diff
@@ -0,0 +1,234 @@
+"""Classes for dealing with HasProperties instances with unique IDs"""
+import uuid
+
+from six import string_types, text_type
+
+from .. import base, basic, handlers, utils
+from ..base.instance import CLASS_TYPES
+
+
+class HasUID(base.HasProperties):
+    """HasUID is a HasProperties class that includes unique ID
+
+    Adding a UID to HasProperties allows serialization of more complex
+    structures, including recursive self-references. They are serialized
+    to a flat dictionary of UID/HasUID key/value pairs.
+    """
+
+    _REGISTRY = dict()
+    _INSTANCES = dict()
+
+    uid = basic.String(
+        'Unique identifier',
+        default=lambda: text_type(uuid.uuid4()),
+    )
+
+    def __init__(self, **kwargs):
+        super(HasUID, self).__init__(**kwargs)
+        self._INSTANCES[self.uid] = self
+
+    @handlers.validator('uid')
+    def _ensure_unique(self, change):
+        if self.uid == change['value']:
+            pass
+        elif change['value'] in self._INSTANCES:
+            raise utils.ValidationError(
+                message='Uid already used: {}'.format(change['value']),
+                reason='invalid',
+                prop=change['name'],
+                instance=self,
+            )
+        return True
+
+    @handlers.observer('uid')
+    def _update_instances(self, change):
+        self._INSTANCES.update({change['value']: self})
```

Comment:
Should we delete old references?

Reply:
That has problems when referring to the instance by the uid. It's easy to delete old references in `_INSTANCES` but harder to update every reference to the instance. Possible...? Maybe worth it...? But easier just to leave them hanging around.

I guess it's not great if you want to create a new instance that reuses an old uid...